### ISSUE TITLE
Bundler dependency confusion issues not resolved?

### ISSUE THREAD

**Issue by pvalena** [rubygems/rubygems#4694]
### Describe the problem as clearly as you can

Following up on my successful test backport (#4687), I've tried to run the test[*] with Bundler 2.2.20 and Bundler 2.2.21 (but maybe in not in a very smart way). The test fails, although it was supposed to be fixed.

This was prompted with a failing shell [reproducer](https://gist.github.com/pvalena/d6113818216902ea1c5f3d428928df66), from one of my collegues, which AFAICT is doing exactly the same thing, as in the Bundler test, but it is failing with Bundler 2.2.18+.

[*] Test name: bundle install with gems on multiple sources with source affinity when a top-level gem has an indirect dependency and the dependency is in both the top-level and a pinned source installs the dependency from the top-level source without warning

### Post steps to reproduce the problem

I've taken Ruby 2.7.3 sources, and replaced the Bundler in them with Bundler 2.20.21. Then I've run the backported test: 

https://gist.github.com/08e9423766b83a8fb0e5e1a64fa4ca74

### Additional

Is there something I'm doing wrong, or is the test really failing?

I've run the test test  from rubygems git repository as well, and it did succeed (with Ruby 2.7), although I had to do a [workaround](https://git.io/JnHHo) to make it work.

**Comment by deivid-rodriguez** [comment]
I'll investigate tomorrow. In principle it seems like your reproducer indeed suffers from the issue. I think it's related to the gems being served by servers using the old marshalled index, instead of more modern dependency APIs. In that case, the issue is still present because, indeed.

**Comment by Buju77** [comment]
I'm not sure if my issue is related to yours, but starting with bundler 2.2.18 and up, our `bundle update` suddenly fails, while 2.2.17 works perfectly fine.

we are getting dependency issues like:

```
Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.

Bundler could not find compatible versions for gem "google-api-client":
  In snapshot (Gemfile.lock):
    google-api-client (>= 0.38.0)

  In Gemfile:
    google-api-client (~> 0.38)

    fastlane (~> 2.150) was resolved to 2.181.0, which depends on
      google-api-client (>= 0.37.0, < 0.39.0)
```

All 3 different version requirements looks fine for me and should match 0.38.0. But why does bundler 2.2.18+ now suddenly throw an error?

**Comment by deivid-rodriguez** [comment]
To explain a bit more, it's the following piece of code (the secure branch in the condition is the first one):

https://github.com/rubygems/rubygems/blob/e3e299f10d6468f48a90fb95cb2b0cdbbdf40a33/bundler/lib/bundler/definition.rb#L874-L878

The second branch of the condition is only used when either of these two conditions are not met:

https://github.com/rubygems/rubygems/blob/e3e299f10d6468f48a90fb95cb2b0cdbbdf40a33/bundler/lib/bundler/definition.rb#L521-L523

That's either:

* When any of the involved gem servers does not implement a dependency API, just because I couldn't make it work. This is, I believe, the case you're running into. It doesn't affect rubygems.org or most other custom gem servers out there, but it does affect you toy reproducer.
* For backwards compatibility, when the Gemfile has multiple global sources.

**Comment by pvalena** [comment]
Thanks for explanation!

I've overlooked "When **any** of the involved gem servers does not implement a dependency API", and can confirm the issue persist even when `rubygems.org` is used in combination with `gem server`. This is very unfortunate.
_ _ _ _
```
* Reproducer: reproducer5.Gemfile
Fetching source index from http://127.0.0.1:8802/
Fetching gem metadata from https://rubygems.org/
Fetching source index from https://rubygems.org/
Resolving dependencies...
Using bundler 2.2.21
Fetching a_okay 0.1.0
Installing a_okay 0.1.0
Fetching bar 0.0.1
Installing bar 0.0.1
Bundle updated!
Installed version of bar:
  * bar (0.0.1)
Installed version of a_okay:
  * a_okay (0.1.0)
FAIL - malicious a_okay version installed

$ cat reproducer5.Gemfile
source 'https://rubygems.org'

source "http://127.0.0.1:8802" do
  # requires 'a_okay'
  gem "bar"
end
```
Note that the 'private' (scoped source) repository contains gem 'a_okay' `0.0.1` (not the actual one from rubygems.org). The issue can be worked around when listing `a_okay` in [the scope](https://git.io/JcGGO).

_ _ _ _ 

What comes to my mind, if there's no fix for the insecure branch,  wouldn't it be helpful to warn user they're using an unsafe implementation, or disable it's use altogether, and require them to explicitly enable it?

**Comment by deivid-rodriguez** [comment]
Yep, to be more precise, currently it doesn't work if any of the involved _non global_ gem servers does not implement a dependency API.

The problem is that ir order to figure out which dependencies need to be picked up from the scoped sources, we need to figure out all transitive dependencies of the gems inside the scoped source. In order to achieve that, if we have a dependency API, we can ask the API easily, but if we don't we need to traverse the whole index, we can be painfully slow.

I guess this is not too bad if the index is small, so we could allow that and fix the issue up to a certain index size, and print a warning otherwise?

**Comment by pvalena** [comment]
> I guess this is not too bad if the index is small, so we could allow that and fix the issue up to a certain index size, and print a warning otherwise?

Sounds great!

**Comment by thoger** [comment]
Does use of any git source cause fallback to unsafe resolution even if all configured gem servers provide the dependency API?  Think of setups with a public repo (rubygems.org), private repo, and one or more gems installed from git sources.

**Comment by deivid-rodriguez** [comment]
Nop, `git` sources are not involved here in any way.

**Comment by deivid-rodriguez** [comment]
As far as I know, only dummy gem servers created like in this example are affected. Private gem servers using geminabox or gemstash are ok.
